File: q1625/q1625.java
class Solution {
    public String findLexSmallestString(String s, int a, int b) {
        Set<String> visited = new HashSet<>();
        Stack<String> stack = new Stack<>();
        stack.add(s);
        String minStr = s;

        while (!stack.isEmpty()) {
            String currStr = stack.pop();
            if (visited.contains(currStr)) continue; 
            visited.add(currStr);
            if (currStr.compareTo(minStr) < 0) minStr = currStr;

            String added = add(currStr, a);
            stack.add(added);

            String rotated = rotate(currStr, b);
            stack.add(rotated);
        }

        return minStr;
    }

    private String add(String s, int a) {
        int[] arr = new int[s.length()];
        for (int i = 0; i < s.length(); i++) {
            arr[i] = s.charAt(i) - '0';
        }

        for (int i = 1; i < s.length(); i += 2) {
            arr[i] = (arr[i] + a) % 10;
        }

        StringBuilder sb = new StringBuilder();
        for (int num : arr) sb.append(num);
        return sb.toString();
    }

    private String rotate(String s, int b) {
        String p1 = s.substring(0, b);
        String p2 = s.substring(b, s.length());
        return p2 + p1;
    }
}

File: q1625/q1625.py
class Solution:
    def findLexSmallestString(self, s: str, a: int, b: int) -> str:
        visited = set()
        stack = [s]
        minStr = s
        
        while stack:
            curr = stack.pop()
            if curr in visited:
                continue
            visited.add(curr)
            if curr < minStr:
                minStr = curr

            # Operation 1: Add 'a' to all digits at odd indices
            chars = list(curr)
            for i in range(1, len(s), 2):
                chars[i] = str((int(chars[i]) + a) % 10)
            added = ''.join(chars)
            stack.append(added)

            # Operation 2: Rotate by b
            rotated = curr[-b:] + curr[:-b]
            stack.append(rotated)

        return minStr

